lib/auth/auth_with_roles.go

```diff
@@ -2257,6 +2258,35 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 	}
 	// add implicit roles to the set and build a checker
 	roleSet := services.NewRoleSet(parsedRoles...)
+	dbService := req.RouteToDatabase.ServiceName
+	if dbService != "" {
+		dbServers, err := a.authServer.GetDatabaseServers(ctx, apidefaults.Namespace)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		var database types.Database
+		for _, dbServer := range dbServers {
+			db := dbServer.GetDatabase()
+			if db.GetName() == dbService {
+				database = dbServer.GetDatabase()
+				break
+			}
+		}
+		if database == nil {
+			return nil, trace.NotFound("database %q", dbService)
+		}
+		dbRoute := tlsca.RouteToDatabase{
+			ServiceName: dbService,
+			Protocol:    req.RouteToDatabase.Protocol,
+			Username:    req.RouteToDatabase.Username,
+			Database:    req.RouteToDatabase.Database,
+		}
+		// rough sanity check for db login. Full RBAC checks will ultimately happen on db connect.
+		err = roleSet.CheckAccessToDatabase(dbRoute, database)
```

Comment:
Could you take a look at isMFARequired logic where we already checking access to a db resource by `checker.CheckAccess(db...) : https://github.com/gravitational/teleport/blob/74f2440683a46e703438a1b7e45628242b376902/lib/auth/auth.go#L3289-L3317

Reply:
https://github.com/gravitational/teleport/blob/eb155d120cbc6f368d9c5d486d876fc2fc294d63/lib/auth/auth.go#L3339-L3350

In the code below we see how that `noMFAAccessError` is handled. The comment about masking the access errors is interesting. When tsh checks if MFA is required, if access will be denied then `IsMFARequired` will return false and no other message will be printed. The connection attempt will be denied later.

Anyway, the RBAC checking is done by role.checkAccess function which gives an opaque "access denied" error by design, so as to not act like an oracle. I can't reuse that function by itself for 3 reasons:
1. It's intentionally not exported from the `services` package, and neither is `accessChecker` - so if I move these checks out of RoleSet as Roman suggested, then I can't use these functions.
2. I need the error to be useful client side, but remain opaque server side.
3. Client side (and even in the auth server) we don't have enough info to use that function when --db-user or --db-name is not specified. For example, on the remote db service side the postgres connection is initialized and postgres tells the db service what the default user name is, which is whatever user started the postgres process on the remote host... But client side we just have a blank db_user in our route, so CheckAccess would deny any user which is not allowed either "*" or "" as a db_user.

I think I will take Roman's suggestion to move this code out of RoleSet. I'll put it in tsh instead and do these checks before we request a new db cert and before a connection if we don't need to relogin.